woocommerce/woocommerce
# Improve the handling of the deprecated WC()->api property

### Changes proposed in this Pull Request:

Before version 9.0 the `api` property of the `WooCommerce` class (typically accessed as `WC()->api`) held an instance of the `WC_API` class, defined by the legacy REST API. As of WooCommerce 9.0 this property holds `null` unless the Legacy REST API plugin is installed.

Although `WC_API` basically contains the code that handles the Legacy REST API endpoints, it also has a `get_endpoint_data` method that can be used to invoke the WooCommerce (not only legacy) REST API endpoints from code. In 9.0 this method was moved to a new `RestApiUtil` class, but this is still a breking change for plugins that invoke `WC()->api->get_endpoint_data`, which will break in setups where the Legacy REST API plugin isn't active if they don't get modified to use the method from `RestApiUtil` instead.

To solve this, this plugin:

1. Sets the `WC()->api` property to point to an instance of `LegacyRestApiStub` (which gets a reference to `RestApiUtil` and a facade `get_endpoint_data` method), so that `WC()->api->get_endpoint_data` will continue to work; but **only** if the Legacy REST API plugin is _not_ installed (if it is, `WC()->api` will be an instance of `WC_API` as provided by the plugin).
2. Introduces the `legacy_rest_api_is_available` method in the `WooCommerce` class. This is an exception to our policy of not including any new code in the `includes` classes: this method needs to be prominently discoverable and readily available.
3. Changes the logic to detect if the Legacy REST API plugin is installed from checking if `WC()->api` is null to using `WC()->legacy_rest_api_is_available()`. Note that the logic is inverted: `WC()->legacy_rest_api_is_available()` will return true when `is_null(WC()->api)` would return false, and the other way around.

For the first change to be compatible with the handling performed by the Legacy REST API plugin `WC()->api` is now a private property, and its getting and setting is handled inside the `__get` and `__set` methods of the `WooCommerce` class.

Additionally, if `WC()->api->get_endpoint_data` is invoked without the Legacy REST API plugin active, an error message will be produced (via `wc_doing_it_wrong`): _'WC()->api->get_endpoint_data' is deprecated, please use the following instead: wc_get_container()->get(Automattic\WooCommerce\Utilities\RestApiUtil::class)->get_endpoint_data_.

### How to test the changes in this Pull Request:

1. If you have the Legacy REST API plugin installed, make sure that it's deactivated.

2. Run the following in CLI: `wp eval 'echo get_class(WC()->api);'`, the following should be printed: `Automattic\WooCommerce\Internal\Utilities\LegacyRestApiStub`.

3. Run the following in CLI: `wp eval 'var_dump(WC()->api->get_endpoint_data("/wc/v3"));'`, you should see the full schema of the WooCommerce JSON REST API v3; and if you have `WP_DEBUG` set to true, you'll also see an error message indicating that `WC()->api->get_endpoint_data` is deprecated. 

4. Query any of the Legacy REST API endpoints, for example `/wc-api/v3/orders`; you should get a 404 response with the message _"The WooCommerce API is disabled on this site"_ inside the body.

5. Create a webhook that has "Legacy API (v3)" in the "API Version" field. Verify that the description of the selected value has the "NOT AVAILABLE" warning as part of its name.

6. Run the following in CLI: `wp eval 'WC_Admin_Notices::reset_admin_notices();'`.

7. Reload the webhooks settings page (WooCommerce - Settings - Advanced - Webhooks). You should see an "Incompatible webhooks warning" notice, and an exclamation mark next to the name of the webhook you just created.

9. Load the main WooCommerce page, you should see a "WooCommerce webhooks that use the Legacy REST API are unsupported" notice.

10. Load the Legacy API settings page (WooCommerce - Settings - Advanced - Legacy API). You should see the _"The legacy REST API is NOT enabled"_ as the setting title, and the checkbox should be unchecked.

11. Load the status report page (WooCommerce - Status - System Status), in the "WordPress environment" section the value for "WooCommerce Legacy REST API package" should be "The Legacy REST API plugin is not installed on this site".

12. Now install (if it's not installed already) and activate the Legacy REST API plugin.

13. Repeat step 2, this time you should get `WC_API` and no error message.

14. Repeat step 3, you should get the same result.

15. Repeat step 4, this time you should get the data as expected from the endpoint.

16. Open (or reload) the edit page for the webhook that you created in step 5, this time the description of the value for "API Version" should say "deprecated" instead of "NOT AVAILABLE".

17. Repeat steps 6 to 11, notices should have disappeared and the texts in the setting and the status item should have changed accordingly.

### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
